### PR TITLE
Reimplement Line drawing, implement Triangle object

### PR DESCRIPTION
**Pull request by Byter09**
Hi again.

As I already mentioned in a comment below #279, I hereby implemented the line drawing functionality with some additional features.
These are:
- Point/Line collision
- Line/Line collision
- Line/Rectangle collision
- Line/Circle collision
- a `distance` method for `Vector`
- Lines now have a standard thickness which can be overridden by using `with_thickness`
- thanks to the new `DrawAttributes`, `Line`s can also be transformed now. Thanks again for that!

Things that are still missing:
- more objects (triangles, quadrilaterals, basically any polygons I'm still working on, this includes an overhauled circle) but I wanted to at least restore line drawing
- `Line` does not support the `constrain` method because I honestly have no idea what it does
  - does it fit the object to be in that rectangle?
  - if so, whats the strategy? keep ratio? and how should that work with lines?

I'll continue working on triangles and push them next. Every polygon with more vertices will follow in the next two weeks, depending on whether #264 gets resolved in that time. I highly recommend getting rid of `Shape` (which is still planned in #279, yay), so each `Drawable` has to implement with what it can overlap itself. This is really helpful when the amount of shapes rises.

I hope this PR meets all the standards and is ready to merge into `new-draw-interface` and later `development`.

EDIT: I should probably mention, that I did turn off rustfmt for this PR, as it was an absolute nightmare to look through all the changes.

### CODE REVIEW

**Comment by Byter09** [comment]
As this PR was not yet merged, I took the liberty to add an additional object - the `Triangle`.

The triangle object has its own `overlaps_*` methods, which are mostly based on `Line` overlaps methods, as we only have to check if something crosses its edges. I also updated the geometry example and used a scale transform, as there was no demonstration of it yet.

The triangle now allows to build basically all polygons, including the already existing rectangle and circle, which will be my next step. But, as #279 is not ready for development, I will not work on a replacement but additions with different names, until that is resolved and they may be used instead.

The main point of replacing and using `Triangle`s to draw `Circle`, is the possibility of having n vertices, in which a circle, in relation to its size, can always be drawn smoothly (though, this has to be some optional feature, probably with `Option<u32>`, where `None` will be automatic).

**Comment by ryanisaacg** [comment]
I appreciate the PR!

Question: Why put this in `geom/objects` instead of just `geom`?

Also, there's a slight problem with using a Triangle to draw the existing shapes: index-optimization. Basically, if I draw a rectangle, I am sending 2 triangles to the GPU, but there's only 4 different points we need to draw. So we supply 4 vertices and 6 indices.

If we wanted to replicate this behavior with Triangle, each triangle in a shape would duplicate its indices, which would be a significant memory overhead.

**Comment by Byter09** [comment]
`geom/objects` because `geom` is getting kind oft cluttered. I thought having basic stuff in `geom` and shapes in their own folder would make it a bit cleaner. I also planned to add more, also in that folder. Thanks to Rusts module system, noone will notice :P

Oh and you're right about the index-optimization. I kind of forgot about that. Well... Then let me rephrase: The triangle is an additional shape. Full stop. :D
And regarding the rectangle, it actually can stay as it is, as the next task on my list was the quadrilateral, so drawing trapezes and such is possible. This is a bit harder because the shape can overlap itself and I would have to write a triangulation algorithm (which would probably be the ear cutting algorithm, as it is fast and produces non-overlapping triangles).

**Comment by ryanisaacg** [comment]
I can go for the more advanced objects being in a subfolder. As for triangulation, if you want an example of it being done in quicksilver, the raycast example is making an arbitrarily complex concave polygon and turning it into triangles.

**Comment by Byter09** [comment]
It is, apart from the missing `constrain` methods for `Line` and `Triangle` as I'm having problems to find more information on what exactly they should do. But I guess they're not actively in use and having `unimplemented!()` in dev code can't hurt.

**Comment by ryanisaacg** [comment]
Oh, constrain creates a copy of a shape that's within the outer Rectangle, having moved the minimum distance. It's useful for constraining a camera area to a level border, for example. I think just not having a constrain function for line and triangle makes sense for now.

**Comment by Byter09** [comment]
Alright, that bit I understood, but as mentioned above: what's the resize strategy? Keep ratio? So basically scaling down until it fits if it doesn't?

**Comment by ryanisaacg** [comment]
There isn't a resize strategy actually; currently the constrain functions can do unexpected things if you pass a shape that can't be constrained just by translating.

**Comment by Byter09** [comment]
Which is the case for `Line` for example. I was simply not sure what exactly should happen so I left it out.
If you merge it today/tomorrow, I'll write a keep-ratio-and-shrink-until-it-fits algorithm for all shapes if you want. At least `Line` should be manageable.

**Comment by Byter09** [comment]
Ohhh I'm sorry. I must have misunderstood something. I just now realised the importance of "translate" as you said. So keep the object its original size but move it so its inside, if it doesn't fit, do nothing. Got it.

I thought this was like the opposite of `bounding_box()`, in which you have a rectangle and make the object fit inside by shrinking if necessary.

Got it. Depending on what the status of this PR is tomorrow (it is 3:47am right now), I'll implement something that should work :)

**Comment by ryanisaacg** [comment]
I look forward to it! 😄

**Comment by ryanisaacg** [comment]
@Byter09 I'll merge this into my PR, get those two functions implemented, and then merge the whole new draw system.
